### PR TITLE
Enable setting multiple external networks

### PR DESCRIPTION
**Pull request by rossella**
Enable multiple external network for Neutron.
Some manual steps are still needed and will be documented in the
official documentation.

### CODE REVIEW

**Comment by rossella** [comment]
Tested with OVS, still need to test with Linuxbridge

**Comment by rossella** [comment]
Tested with Linuxbridge too now

File: chef/cookbooks/neutron/recipes/common_agent.rb [hunk 1]
```diff
@@ -304,7 +338,7 @@
         :periodic_fuzzy_delay => 5,
         :external_network_bridge => external_network_bridge,
         :dvr_enabled => neutron[:neutron][:use_dvr],
-        :dvr_mode => node.roles.include?("neutron-network") ? "dvr_snat" : "dvr"
+        :dvr_mode => node.roles.include?("neutron-network") ? "dvr_snat" : "dvr",
```
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: Unneeded change.

**Comment by vuntz** [comment]
Interesting, need to look at it a bit deeper, but first thing I'm wondering is whether we should actually make the default floating network part of this attribute. This would then be the `external_networks` attribute, instead of `additional_external_networks`.

**Comment by vuntz** [comment]
Does this require the additional networks to be defined in network.json? Or how does this work to get information on how to create these networks?

**Comment by rossella** [comment]
@vuntz yes it requires the network to be added manually to network.json . You can find more info about the configuration needed here [1] (especially the last paragraph) .
It's possible to make the default floating network part of this attribute and I thought about it. It requires a previous cleanup since the name of the default floating network is hard coded in various part of the code (from network.json to barclamp neutron post_install_conf.rb) . 

[1] https://docs.google.com/document/d/1Q8Evv5Z6SViWhzEvZOme1xNlP4lv7Vd1owdoAuesxVI/edit?usp=sharing

File: chef/cookbooks/neutron/recipes/common_agent.rb [hunk 2]
```diff
@@ -89,6 +89,7 @@
   end
 end
 
+multiple_external_networks = !neutron[:neutron][:additional_external_networks].empty?  && node.roles.include?("neutron-network")
```
Comment (rsalevsky) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: two spaces between `.empty?  && node.`

File: chef/cookbooks/neutron/recipes/common_agent.rb [hunk 3]
```diff
@@ -132,8 +133,11 @@
 
   # Create the bridges Neutron needs.
   # Usurp config as needed.
-  [ [ "nova_fixed", "fixed" ],
-    [ "public", "public"] ].each do |net|
+  networks = [ [ "nova_fixed", "fixed" ], ["public", "public"]]
+  neutron[:neutron][:additional_external_networks].each do |net|
+    networks << [ net, net ]
```
Comment (rsalevsky) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: should be `[net, net]`

File: chef/cookbooks/neutron/recipes/common_agent.rb [hunk 4]
```diff
@@ -132,8 +133,11 @@
 
   # Create the bridges Neutron needs.
   # Usurp config as needed.
-  [ [ "nova_fixed", "fixed" ],
-    [ "public", "public"] ].each do |net|
+  networks = [ [ "nova_fixed", "fixed" ], ["public", "public"]]
```
Comment (rsalevsky) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: should `[[`

File: chef/cookbooks/neutron/recipes/common_agent.rb [hunk 5]
```diff
@@ -206,14 +214,27 @@
...
     interface_driver = "neutron.agent.linux.interface.OVSInterfaceDriver"
-    external_network_bridge = "br-public"
+    if !multiple_external_networks
+      external_network_bridge = "br-public"
+      if ml2_type_drivers.include?("vlan")
+        bridge_mappings = "physnet1:br-fixed"
+      end
+    else
+      bridge_mappings = "floating:br-public, "
+      bridge_mappings += neutron[:neutron][:additional_external_networks].collect {|n| n + ":" + "br-" + n }.join ','
```
Comment (rsalevsky) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: space missing after `{`

File: chef/cookbooks/neutron/recipes/common_agent.rb [hunk 6]
```diff
@@ -206,14 +214,27 @@
   ml2_mech_drivers = neutron[:neutron][:ml2_mechanism_drivers]
   ml2_type_drivers = neutron[:neutron][:ml2_type_drivers]
 
+
```
Comment (rsalevsky) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: empty line

**Comment by mjura** [comment]
For me looks OK +1

I wondering if want to add these new external networks and create for them some additional routers and networks in neutron using `post_install_conf.rb` or should we leave it for user.

File: chef/data_bags/crowbar/bc-template-neutron.json [hunk 7]
```diff
@@ -66,7 +67,7 @@
     "neutron": {
       "crowbar-revision": 0,
       "crowbar-applied": false,
-      "schema-revision": 40,
+      "schema-revision": 41,
```
Comment (rhafer) [inline chef/data_bags/crowbar/bc-template-neutron.json]: Revision 41 was already taken by different pull request please take 42 (and rename 041_additional_external_networks.rb accordingly)

**Comment by vuntz** [comment]
Shouldn't we integrate the `crowbar network enable_interface` bit from the doc in the rails app, when applying the barclamp?

**Comment by rhafer** [comment]
> Shouldn't we integrate the crowbar network enable_interface bit from the doc in the rails app, when applying the barclamp?

Yeah. I'd like to see that as well. Should be pretty simple.

File: chef/cookbooks/neutron/recipes/common_agent.rb [hunk 8]
```diff
@@ -221,6 +242,18 @@
     agent_config_path = "/etc/neutron/plugins/linuxbridge/linuxbridge_conf.ini"
     interface_driver = "neutron.agent.linux.interface.BridgeInterfaceDriver"
     external_network_bridge = ""
+    physnet = node[:crowbar_wall][:network][:nets][:nova_fixed].first rescue nil
+    interface_mappings = "physnet1:" + physnet
+    if multiple_external_networks
+      neutron[:neutron][:additional_external_networks].each do |net|
+        n = ":"+ net
```
Comment (rhafer) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: What is this for? ```n``` doesn't seem to be used anywhere.

File: chef/cookbooks/neutron/recipes/common_agent.rb [hunk 9]
```diff
@@ -206,14 +214,27 @@
   ml2_mech_drivers = neutron[:neutron][:ml2_mechanism_drivers]
   ml2_type_drivers = neutron[:neutron][:ml2_type_drivers]
 
+
   case
   when ml2_mech_drivers.include?("openvswitch")
-    # package is already installed
 
+    # package is already installed
```
Comment (rhafer) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: Nitpick: Unneeded change.

File: chef/cookbooks/neutron/recipes/common_agent.rb [hunk 10]
```diff
@@ -206,14 +214,27 @@
...
     neutron_agent = node[:neutron][:platform][:ovs_agent_name]
     agent_config_path = "/etc/neutron/plugins/openvswitch/ovs_neutron_plugin.ini"
     interface_driver = "neutron.agent.linux.interface.OVSInterfaceDriver"
-    external_network_bridge = "br-public"
+    if !multiple_external_networks
```
Comment (rhafer) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: Do we really need to distinguish the multiple vs single external network cases here? Can't we always use the code path that's now in the ```else``` branch?

**Comment by rhafer** [comment]
> > Shouldn't we integrate the crowbar network enable_interface bit from the doc in the rails app, when applying the barclamp?
> 
> Yeah. I'd like to see that as well. Should be pretty simple. (But it would require a similar change to the nova barclamp as well)


@rossella  I just submitted a pull request to your branch that should do the trick.

File: chef/cookbooks/neutron/recipes/common_agent.rb (hunk 10 above)
Comment (rossella) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: I can refactor a bit, you are right. I still need and if/else block because external_network_bridge should not be set if multiple external networks are used

File: chef/cookbooks/neutron/recipes/common_agent.rb (hunk 8 above)
Comment (rossella) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: sorry, leftover from a previous version

**Comment by rossella** [comment]
@rhafer thanks  a lot!

**Comment by rossella** [comment]
@mjura some manual steps are still required to set up multiple networks so I think it's OK to leave the task of creating the router and the external network to the user for now.

**Comment by vuntz** [comment]
> @rossella I just submitted a pull request to your branch that should do the trick.

For reference, https://github.com/rossella/barclamp-neutron/pull/1

> @mjura some manual steps are still required to set up multiple networks

@rossella which manual steps are still required? I think I'd like to see the openstack resources created automatically also (but could be in a later PR, of course).

(This PR will be a great change, btw!)

File: chef/cookbooks/neutron/recipes/common_agent.rb (hunk 10 above)
Comment (rhafer) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: Is ```external_network_bridge``` really required in the case when there is just a single external network? Or can we do without it in that case as well?

Comment (rossella) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: As long as the bridge_mappings is set it's fine if external_network_bridge is not set...I removed external_network_bridge completely and set the bridge_mappings all the time

**Comment by rossella** [comment]
@vuntz the manual step required is to modify /etc/crowbar/network.json to add the additional external networks. Is there an easy way to automate that?

**Comment by rossella** [comment]
@mjura @vuntz I'd prefer to automate the creation of the OpenStack resources in another PR...the post install script needs some refactoring and this PR is getting quite big already

File: chef/cookbooks/neutron/recipes/common_agent.rb [hunk 11]
```diff
@@ -185,7 +189,11 @@
   # Note: As moving between network plugins is currently not supported by this
   #       cookbook this code is mostly just sitting here and waiting for the
   #       plugin switching support to be implemented.
-  [ "br-public", "br-fixed" ].each do |name|
+  bridges = ["br-public", "br_fixed"]
```
Comment (dirkmueller) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: why did you change from br-fixed to br_fixed ?

File: chef/cookbooks/neutron/recipes/common_agent.rb [hunk 12]
```diff
@@ -209,18 +217,34 @@
...
+    end
   when ml2_mech_drivers.include?("linuxbridge")
     package node[:neutron][:platform][:lb_agent_pkg]
 
     neutron_agent = node[:neutron][:platform][:lb_agent_name]
     agent_config_path = "/etc/neutron/plugins/linuxbridge/linuxbridge_conf.ini"
     interface_driver = "neutron.agent.linux.interface.BridgeInterfaceDriver"
-    external_network_bridge = ""
+    physnet = node[:crowbar_wall][:network][:nets][:nova_fixed].first rescue nil
+    interface_mappings = "physnet1:" + physnet
```
Comment (dirkmueller) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: the rescue nil and then +physnet in the line below doesn't look good to me. if this really can not happen, then remove the rescue nil (avoid hiding the real error). if it can happen, the rescue should return something more sensible.

File: chef/cookbooks/neutron/templates/default/l3_agent.ini.erb [hunk 13]
```diff
@@ -50,7 +50,6 @@ handle_internal_only_routers = <%= @handle_internal_only_routers %>
 # empty value for the linux bridge. when this parameter is set, each L3 agent
 # can be associated with no more than one external network.
 # external_network_bridge = br-ex
-external_network_bridge = <%= @external_network_bridge %>
```
Comment (dirkmueller) [inline chef/cookbooks/neutron/templates/default/l3_agent.ini.erb]: hmm, why does the default value of 'br-ex' work?

File: chef/cookbooks/neutron/templates/default/linuxbridge_conf.ini.erb [hunk 14]
```diff
@@ -34,7 +34,7 @@
 #
 # physical_interface_mappings =
 # Example: physical_interface_mappings = physnet1:eth1
-physical_interface_mappings = physnet1:<%=@physnet%>
+physical_interface_mappings = <%=@interface_mappings%>
```
Comment (dirkmueller) [inline chef/cookbooks/neutron/templates/default/linuxbridge_conf.ini.erb]: suggest additional spacing, e.g.

```ruby
   <%= @interface_mappings -%>
```

File: chef/cookbooks/neutron/templates/default/l3_agent.ini.erb (hunk 13 above)
Comment (rossella) [inline chef/cookbooks/neutron/templates/default/l3_agent.ini.erb]: I don't think it works, it's commented...am I missing something?

File: chef/cookbooks/neutron/recipes/common_agent.rb (hunk 11 above)
Comment (rossella) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: ooops

File: chef/cookbooks/neutron/recipes/common_agent.rb (hunk 12 above)
Comment (rossella) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: rescue nil removed

File: chef/cookbooks/neutron/templates/default/l3_agent.ini.erb (hunk 13 above)
Comment (mjura) [inline chef/cookbooks/neutron/templates/default/l3_agent.ini.erb]: For me it is ok look comment above `when this parameter is set, each L3 agent can be associated with no more than one external network`

File: crowbar_framework/app/models/neutron_service.rb [hunk 15]
```diff
@@ -277,6 +299,9 @@ def apply_role_pre_chef_call(old_role, role, all_nodes)
 
     network_nodes.each do |n|
       net_svc.allocate_ip "default", "public", "host",n
```
Comment (mjura) [inline crowbar_framework/app/models/neutron_service.rb]: `net_svc.allocate_ip "default", "public", "host",n`

can be removed, look on comment from:
https://github.com/crowbar/barclamp-neutron/pull/227/files#diff-1cab7f3b6842279ae4d6bd875e073cacL279

File: chef/cookbooks/neutron/templates/default/l3_agent.ini.erb (hunk 13 above)
Comment (dirkmueller) [inline chef/cookbooks/neutron/templates/default/l3_agent.ini.erb]: then we should imho explicitely set external_network_bridge to empty

File: crowbar_framework/app/models/neutron_service.rb (hunk 15 above)
Comment (rhafer) [inline crowbar_framework/app/models/neutron_service.rb]: > net_svc.allocate_ip "default", "public", "host",n
> 
> can be removed,

That's not really true. It could be changed to an appropriate enable_interface call. But let's worry about that in the other pull request (#227).

**Comment by mjura** [comment]
+1 if you will fix comment from @dirkmueller

**Comment by dirkmueller** [comment]
LGTM

**Comment by vuntz** [comment]
For the record: this seems to break ovs support.
